lib/schedule/block.ex

```diff
@@ -40,28 +40,56 @@ defmodule Schedule.Block do
 
   @spec blocks_from_trips([Trip.t()]) :: by_id()
   def blocks_from_trips(trips) do
-    trips
-    |> group_trips_by_block()
-    |> Helpers.map_values(&block_from_trips/1)
-  end
+    {revenue_trips, nonrevenue_trips} =
+      Enum.split_with(trips, fn trip -> trip.route_id != nil end)
+
+    nonrevenue_trips =
+      Enum.group_by(nonrevenue_trips, fn trip -> {trip.block_id, trip.schedule_id} end)
 
-  @spec group_trips_by_block([Trip.t()]) :: %{key() => [Trip.t()]}
-  defp group_trips_by_block(trips) do
-    trips
+    revenue_trips
     |> Enum.filter(fn trip -> trip.stop_times != [] end)
     |> Enum.group_by(fn trip -> {trip.block_id, trip.service_id} end)
-    |> Helpers.map_values(&sort_trips_by_time/1)
+    |> Helpers.map_values(fn trips ->
+      block_id = List.first(trips).block_id
+      schedule_id = List.first(trips).schedule_id
+      deadheads = nonrevenue_trips[{block_id, schedule_id}] || []
+      block_from_trips(trips, deadheads)
+    end)
   end
 
-  @spec block_from_trips([Trip.t()]) :: t()
-  def block_from_trips([first_trip | _] = trips) do
+  @spec block_from_trips([Trip.t()], [Trip.t()]) :: t()
+  def block_from_trips(revenue_trips, nonrevenue_trips \\ []) do
+    revenue_trips = Enum.sort_by(revenue_trips, & &1.start_time)
+    nonrevenue_trips = Enum.sort_by(nonrevenue_trips, & &1.start_time)
+    [first_trip | _] = revenue_trips
```

Comment:
This is a line that I think is clearer _not_ using pattern matching: `first_trip = List.first(revenue_trips)`.